src/catalog/src/durable/impls/persist.rs

```diff
@@ -239,18 +276,37 @@ impl UnopenedPersistCatalogState {
             }
         }
 
-        UnopenedPersistCatalogState {
+        Ok(UnopenedPersistCatalogState {
             since_handle,
             write_handle,
             listen,
             persist_client,
-            shard_id,
+            shard_id: catalog_shard_id,
             snapshot,
             upper,
             epoch,
             configs,
+            organization_id,
             metrics,
-        }
+        })
+    }
+
+    async fn upgrade_version(
+        persist_client: &PersistClient,
+        upgrade_shard_id: ShardId,
+    ) -> Option<semver::Version> {
```

Comment:
When would this return a None? When the upgrade shard hasn't been created yet? Might be simpler to reason about if we create it on demand and always return a version?

Reply:
> When would this return a None? When the upgrade shard hasn't been created yet? 

Yep exactly, when the shard hasn't been created yet.

> Might be simpler to reason about if we create it on demand and always return a version?

Yes I agree, how do I go about creating a shard in persist? Just open a writer and it will get created?

Reply:
Yup open a writer or a reader would do it

Reply:
I might be misunderstanding, but I don't think this would work. Opening a writer increments the version of the data shard. So we need to first check the version, before opening a writing to guarantee that a version exists to know if opening a writer is safe. When we check the version, we need to deal with a version not existing because it's never been opened before.

Reply:
I'm probably just missing something. I have to run a bit early today (now), wanna huddle on this real quick first thing tomorrow?

Reply:
Yep, sound good.

Reply:
Oh I think I figured out what I was missing. What if we instead did something like (slight handwave here on the edge cases) `.unwrap_or(last_version_without_this_PR_in_it)`?